### PR TITLE
Adding docs build env for osx

### LINKED ISSUE #10026: Cannot build docs from current master
### Description

Cannot build docs from current master.

### Expected behavior

`tox -e build_docs` should build the documentation.

### Actual behavior

I got instead the error below and no output `docs/_build/html/index.html` file.
```
...
reading sources... [ 99%] whatsnew/4.1
reading sources... [100%] whatsnew/index


Warning, treated as error:
/Users/aldcroft/git/astropy/docs/units/index.rst.rst:402:autosummary: stub file not found 'astropy.units.function.logarithmic.m_bol'. Check your autosummary_generate setting.
ERROR: InvocationError for command /Users/aldcroft/git/astropy/.tox/build_docs/bin/sphinx-build -W -b html . _build/html (exited with code 2)
___________________________________________________________________________ summary ____________________________________________________________________________
ERROR:   build_docs: commands failed
```

### Steps to Reproduce

```
git checkout master
git pull upstream master
git clean -fxd
rm -rf .tox/build_docs
tox -e build_docs
```

```python
# Put your Python code snippet here.
```

### System Details

```
Darwin-19.2.0-x86_64-i386-64bit
Python 3.6.10 |Anaconda, Inc.| (default, Jan  7 2020, 15:01:53) 
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)]
Numpy 1.18.1
astropy 4.1.dev737+gade7e2492
Scipy 1.4.1
```

### PR DESCRIPTION
**Pull request by bsipocz**
osx is case insensitive thus we always get warnings for ``m_bol`` and ``M_bol``. This PR thus adds a new tox env to handle it.

Alternative: create a `build_docs_strict` that has the `-W` and being the one CI runs. pro: more convenient for osx users, con: folks will be surprised why the docs build works for them locally while it fails on CI.

Close #10026

### CODE REVIEW

**Comment by pllim** [comment]
Shouldn't we let the CI run?

**Comment by bsipocz** [comment]
The new env is not used in CI, so I don't think it's needed. What it needs is whether @taldcroft is happy with the workaround, or we should do the alternative instead.

**Comment by Cadair** [comment]
This isn't really osx specific in anyway, so I would say it at least needs a better name. Also there is definitely ways to write this with less duplication, but not that important.

**Comment by bsipocz** [comment]
I'm happy with if you push a commit that removes the duplication. However, on the name I'm mostly interested in @taldcroft's and other osx user's ideas as this only affect us, and we need a name that serves the best this crowd. (as neither `withoutwarnings` or `non_strict`, etc, is not descriptive enough of the fact that the docs will never build with `-W` on osx as long as we have both those case sensitive units.

**Comment by taldcroft** [comment]
From my perspective this is what I would suggest for naming:

- `build_docs` would be without -W.  Like most developers I want to build the docs and see how many sphinx mistakes I made and *also* look at the rendered docs at that point.
- `test_build_docs` would be with -W and aimed mostly at CI.  No joy unless it is perfect.

My OCD is singing a little at this point because every other tox option uses dashes and these two use underscores.  By precedent it would be `test-build_docs` but that is even worse.

BTW, Windows is not case sensitive.  There are not a ton of devs on Windows AFAIK, but I wonder if the same fail happens there for building docs, or if it is even possible at all given other dependencies.

**Comment by Cadair** [comment]
Another option is we make `build_docs` not have the `-W` and just configure the CI to pass that flag, i.e `tox -e build_docs -- -W` it seems overkill to have a whole env in the tox config for `-W`

**Comment by pllim** [comment]
Re: Windows -- I just don't build docs on Windows. I build it on Linux. 😉

**Comment by taldcroft** [comment]
:+1: on the idea from @Cadair

**Comment by bsipocz** [comment]
I don't have strong feelings but neither time, so any can take over this PR as they wish (I just wanted to close up an issue with a few liner). 
Originally we indeed agreed to do the opt-in `-W` for the CI, but then I foresee similar amount of complaints and questions to flake8 issues of why the build is failing when it passed locally etc.

**Comment by taldcroft** [comment]
I'm also completely fine with your original alternative of `build_docs_strict`, but just giving it the name `test_build_docs` to be generally more consistent with the rest.  As you say this gives an easy way for developers to either (a) get the docs built locally or (b) test that it will pass on CI.

**Comment by bsipocz** [comment]
yes, `test_build_docs` is a good/better name indeed.

**Comment by astrofrog** [comment]
FWIW I prefer ``build_docs-strict`` since it's then easy to conditionally check for the presence of 'strict' and add the flag rather than make a whole new configuration.

Also I don't like having ``_`` in ``build_docs`` - I originally had just ``docs`` but people wanted it to look like the ``setup.py`` command. ``build-docs`` is not really an option since ``-`` has special meaning in tox.

**Comment by astrofrog** [comment]
To be completely honest, we really should just find a way to fix that bug, e.g. by identifying ambiguous cases like Mbol/mbol and coming up with some kind of convention for that :grin:

**Comment by bsipocz** [comment]
Oh, fixing the issue would be totally awesome solution. It just didn't seem there was a way and will to do it.

**Comment by taldcroft** [comment]
> To be completely honest, we really should just find a way to fix that bug

Well, sort of. Probably 90% of the time when I do substantial doc edits, there is some warning from a mistake I made, so in the current config you get an error from tox and don't get the chance to look at rendered HTML.  In fact it is *much worse*.  After waiting several minutes after `tox -e build_docs` with reading all the sources, I finally get ...
```
reading sources... [ 99%] whatsnew/4.0
reading sources... [ 99%] whatsnew/4.1
reading sources... [100%] whatsnew/index

Warning, treated as error:
/Users/aldcroft/git/astropy/docs/units/index.rst.rst:402:autosummary: stub file not found 'astropy.units.function.logarithmic.m_bol'. Check your autosummary_generate setting.
ERROR: InvocationError for command /Users/aldcroft/git/astropy/.tox/build_docs/bin/sphinx-build -W -b html . _build/html (exited with code 2)
build_docs: Shutting down tox-pypi-filter server
_____________________________________________________________________ summary ______________________________________________________________________
ERROR:   build_docs: commands failed
```
So that means a turnaround of about 5 minutes to find just the *first* sphinx error.  Yck.

Basically I'm not using tox in part because of issues like this, and I'm sure that wasn't the intent of APE-17.  (And because `pytest astropy/table -k blah --pdb` is just so much faster and obvious).

**Comment by taldcroft** [comment]
Somewhat related, we often mark documentation issues as good-first-issue or easy, but when I think about it actually successfully building the docs is one of the hardest things (at least for me).  Having a simple `tox -e build_docs` that actually always works would lower that bar a bit.

**Comment by taldcroft** [comment]
About the name, I don't care too much as long as there is a documented way to build the docs that reliably gets to the end of the process of building the docs.

**Comment by mhvk** [comment]
How about going to `tox -e docs` for the non-warning-is-error case and `build_docs` for the one with warnings? Building suggests it is for use outside of one's session... And fine if `build_docs` is a synonym for `test-docs` (which I rather like as being a strict test of whether the docs are OK).

**Comment by astrofrog** [comment]
If we want ``docs`` to be the non-warning one, we could also make it be parallel by default and I think it might be possible to make it just use ``pip install -e .`` for the install which would speed things up a lot.

**Comment by bsipocz** [comment]
I like this latest suggestion. 

Feel free to push commits here, or close this off as it's quicker than type the comment and rely on me understanding what  was  proposed exactly.

**Comment by astrofrog** [comment]
I can try and push a commit this week

**Comment by mhvk** [comment]
@bsipocz - bit rude to change your milestone perhaps, but as I'm going through anyway, it does seem this is not essential for 4.1 as such. (here's to hoping my "helping" actually helps...)

**Comment by taldcroft** [comment]
@bsipocz and others -- I just tried this but with the tiny changes suggested in previous comments:
- Change `build_docs_osx` to `docs`
- Change `pip freeze` to `pip install -e ..` (since this runs in `./docs`).

This worked great for me and I got a docs build from tox!  First time ever that I saw the congratulations.
```
__________________________________________________________________________ summary ___________________________________________________________________________
  docs: commands succeeded
  congratulations :)
```
So I suggest to not let perfect be the enemy of good, reopen this and do those little changes. (And I guess some narrative docs update for this new process for docs building).  I'm happy to do it all, but don't want to steal your PR.

**Comment by pllim** [comment]
> I'm happy to do it all, but don't want to steal your PR.

As long as you pull this branch down and cherry pick that hash, I think she would get the credit anyway.

**Comment by astrofrog** [comment]
@taldcroft @bsipocz - I would recommend to whoever opens the PR to then also add ``-j auto`` to the ``sphinx-build`` command for the ``docs`` env, which will enable fast parallel builds.

@taldcroft - instead of changing the pip command, one should add ``usedevelop=true`` to that environment, which will make tox install the package in editable mode before building the docs.

**Comment by taldcroft** [comment]
Okay, will do. As @bsipocz noted this PR is editable by maintainers and so I'll just try to finish it up.

**Comment by saimn** [comment]
See #10771.

**Comment by saimn** [comment]
I think this one can be close since the documentation build was moved to RTD and does not use tox, and the `-W` option in the tox `build_docs` env was removed in #10497.

**Comment by pllim** [comment]
I thought this was mostly for Mac developers who want to build doc locally with tox?

**Comment by saimn** [comment]
The `m_bol` vs `M_bol` warning is Mac specific (or actually case-insensitive filesystem specific), but the idea to remove `-W` applies to everyone I think, since the tox env is now only for users.
